File: public/downloadables/croissant_api.py
import requests
from dataclasses import dataclass, field
from typing import List, Optional, Dict, Any

CROISSANT_BASE_URL = 'https://croissant-api.fr/api'

@dataclass
class Game:
    gameId: str
    name: str
    description: str
    price: float
    owner_id: str
    showInStore: bool
    iconHash: Optional[str] = None
    splashHash: Optional[str] = None
    bannerHash: Optional[str] = None
    genre: Optional[str] = None
    release_date: Optional[str] = None
    developer: Optional[str] = None
    publisher: Optional[str] = None
    platforms: Optional[List[str]] = None
    rating: float = 0.0
    website: Optional[str] = None
    trailer_link: Optional[str] = None
    multiplayer: bool = False
    download_link: Optional[str] = None

@dataclass
class User:
    userId: str
    username: str
    email: Optional[str] = None
    verified: bool = False
    studios: Optional[List[dict]] = None
    roles: Optional[List[str]] = None
    inventory: Optional[List[dict]] = None
    ownedItems: Optional[List[dict]] = None
    createdGames: Optional[List[Game]] = None
    verificationKey: Optional[str] = None
    steam_id: Optional[str] = None
    steam_username: Optional[str] = None
    steam_avatar_url: Optional[str] = None
    isStudio: Optional[bool] = None
    admin: Optional[bool] = None
    disabled: Optional[bool] = None
    google_id: Optional[str] = None
    discord_id: Optional[str] = None
    balance: Optional[float] = None
    haveAuthenticator: Optional[bool] = None

@dataclass
class Item:
    itemId: str
    name: str
    description: str
    owner: str
    price: float
    iconHash: str
    showInStore: Optional[bool] = None
    deleted: Optional[bool] = None

@dataclass
class InventoryItem:
    user_id: Optional[str] = None
    item_id: Optional[str] = None
    amount: int = 0
    metadata: Optional[Dict[str, Any]] = None
    itemId: str = ''
    name: str = ''
    description: str = ''
    iconHash: str = ''
    price: float = 0.0
    owner: str = ''
    showInStore: bool = False

@dataclass
class Studio:
    user_id: str
    username: str
    verified: bool
    admin_id: str
    isAdmin: Optional[bool] = None
    apiKey: Optional[str] = None
    users: Optional[List[dict]] = None

@dataclass
class Lobby:
    lobbyId: str
    users: List[dict]

@dataclass
class TradeItem:
    itemId: str
    amount: int
    metadata: Optional[Dict[str, Any]] = None

@dataclass
class Trade:
    id: str
    fromUserId: str
    toUserId: str
    fromUserItems: List[dict]
    toUserItems: List[dict]
    approvedFromUser: bool
    approvedToUser: bool
    status: str
    createdAt: str
    updatedAt: str

@dataclass
class OAuth2App:
    client_id: str
    client_secret: str
    name: str
    redirect_urls: List[str]

class CroissantAPI:
    def __init__(self, token: Optional[str] = None):
        self.token = token

    def _headers(self, content_type: bool = False):
        headers = {}
        if self.token:
            headers['Authorization'] = f'Bearer {self.token}'
        if content_type:
            headers['Content-Type'] = 'application/json'
        return headers

    # --- USERS ---
    def get_me(self) -> User:
        if not self.token:
            raise Exception('Token is required')
        r = requests.get(f'{CROISSANT_BASE_URL}/users/@me', headers=self._headers())
        r.raise_for_status()
        return User(**r.json())

    def search_users(self, query: str) -> List[User]:
        r = requests.get(f'{CROISSANT_BASE_URL}/users/search', params={'q': query})
        if not r.ok:
            return []
        return [User(**u) for u in r.json()]

    def get_user(self, user_id: str) -> User:
        r = requests.get(f'{CROISSANT_BASE_URL}/users/{user_id}')
        r.raise_for_status()
        return User(**r.json())

    def transfer_credits(self, target_user_id: str, amount: float) -> dict:
        if not self.token:
            raise Exception('Token is required')
        r = requests.post(f'{CROISSANT_BASE_URL}/users/transfer-credits',
                         headers=self._headers(True),
                         json={'targetUserId': target_user_id, 'amount': amount})
        r.raise_for_status()
        return r.json()

    def verify_user(self, user_id: str, verification_key: str) -> dict:
        r = requests.post(f'{CROISSANT_BASE_URL}/users/auth-verification',
                         headers=self._headers(True),
                         json={'userId': user_id, 'verificationKey': verification_key})
        if not r.ok:
            return {'success': False}
        return r.json()

    # --- GAMES ---
    def list_games(self) -> List[Game]:
        r = requests.get(f'{CROISSANT_BASE_URL}/games')
        if not r.ok:
            return []
        return [Game(**g) for g in r.json()]

    def search_games(self, query: str) -> List[Game]:
        r = requests.get(f'{CROISSANT_BASE_URL}/games/search', params={'q': query})
        if not r.ok:
            return []
        return [Game(**g) for g in r.json()]

    def get_my_created_games(self) -> List[Game]:
        if not self.token:
            raise Exception('Token is required')
        r = requests.get(f'{CROISSANT_BASE_URL}/games/@mine', headers=self._headers())
        if not r.ok:
            return []
        return [Game(**g) for g in r.json()]

    def get_my_owned_games(self) -> List[Game]:
        if not self.token:
            raise Exception('Token is required')
        r = requests.get(f'{CROISSANT_BASE_URL}/games/list/@me', headers=self._headers())
        if not r.ok:
            return []
        return [Game(**g) for g in r.json()]

    def get_game(self, game_id: str) -> Game:
        r = requests.get(f'{CROISSANT_BASE_URL}/games/{game_id}')
        r.raise_for_status()
        return Game(**r.json())

    # --- INVENTORY ---
    def get_my_inventory(self) -> dict:
        if not self.token:
            raise Exception('Token is required')
        r = requests.get(f'{CROISSANT_BASE_URL}/inventory/@me', headers=self._headers())
        r.raise_for_status()
        return r.json()

    def get_inventory(self, user_id: str) -> dict:
        r = requests.get(f'{CROISSANT_BASE_URL}/inventory/{user_id}')
        r.raise_for_status()
        return r.json()

    # --- ITEMS ---
    def list_items(self) -> List[Item]:
        r = requests.get(f'{CROISSANT_BASE_URL}/items')
        if not r.ok:
            return []
        return [Item(**i) for i in r.json()]

    def get_my_items(self) -> List[Item]:
        if not self.token:
            raise Exception('Token is required')
        r = requests.get(f'{CROISSANT_BASE_URL}/items/@mine', headers=self._headers())
        if not r.ok:
            return []
        return [Item(**i) for i in r.json()]

    def search_items(self, query: str) -> List[Item]:
        r = requests.get(f'{CROISSANT_BASE_URL}/items/search', params={'q': query})
        if not r.ok:
            return []
        return [Item(**i) for i in r.json()]

    def get_item(self, item_id: str) -> Item:
        r = requests.get(f'{CROISSANT_BASE_URL}/items/{item_id}')
        r.raise_for_status()
        return Item(**r.json())

    def create_item(self, item_data: dict) -> dict:
        if not self.token:
            raise Exception('Token is required')
        r = requests.post(f'{CROISSANT_BASE_URL}/items/create', headers=self._headers(True), json=item_data)
        r.raise_for_status()
        return r.json()

    def update_item(self, item_id: str, item_data: dict) -> dict:
        if not self.token:
            raise Exception('Token is required')
        r = requests.put(f'{CROISSANT_BASE_URL}/items/update/{item_id}', headers=self._headers(True), json=item_data)
        r.raise_for_status()
        return r.json()

    def delete_item(self, item_id: str) -> dict:
        if not self.token:
            raise Exception('Token is required')
        r = requests.delete(f'{CROISSANT_BASE_URL}/items/delete/{item_id}', headers=self._headers())
        r.raise_for_status()
        return r.json()

    def buy_item(self, item_id: str, amount: int) -> dict:
        if not self.token:
            raise Exception('Token is required')
        r = requests.post(f'{CROISSANT_BASE_URL}/items/buy/{item_id}', headers=self._headers(True), json={'amount': amount})
        r.raise_for_status()
        return r.json()

    def sell_item(self, item_id: str, amount: int) -> dict:
        if not self.token:
            raise Exception('Token is required')
        r = requests.post(f'{CROISSANT_BASE_URL}/items/sell/{item_id}', headers=self._headers(True), json={'amount': amount})
        r.raise_for_status()
        return r.json()

    def give_item(self, item_id: str, amount: int, user_id: str, metadata: Optional[dict] = None) -> dict:
        if not self.token:
            raise Exception('Token is required')
        body = {'amount': amount, 'userId': user_id}
        if metadata:
            body['metadata'] = metadata
        r = requests.post(f'{CROISSANT_BASE_URL}/items/give/{item_id}', headers=self._headers(True), json=body)
        r.raise_for_status()
        return r.json()

    def consume_item(self, item_id: str, params: dict) -> dict:
        if not self.token:
            raise Exception('Token is required')
        r = requests.post(f'{CROISSANT_BASE_URL}/items/consume/{item_id}', headers=self._headers(True), json=params)
        r.raise_for_status()
        return r.json()

    def update_item_metadata(self, item_id: str, unique_id: str, metadata: dict) -> dict:
        if not self.token:
            raise Exception('Token is required')
        r = requests.put(f'{CROISSANT_BASE_URL}/items/update-metadata/{item_id}', headers=self._headers(True), json={'uniqueId': unique_id, 'metadata': metadata})
        r.raise_for_status()
        return r.json()

    def drop_item(self, item_id: str, params: dict) -> dict:
        if not self.token:
            raise Exception('Token is required')
        r = requests.post(f'{CROISSANT_BASE_URL}/items/drop/{item_id}', headers=self._headers(True), json=params)
        r.raise_for_status()
        return r.json()

    # --- LOBBIES ---
    def create_lobby(self) -> dict:
        if not self.token:
            raise Exception('Token is required')
        r = requests.post(f'{CROISSANT_BASE_URL}/lobbies', headers=self._headers())
        r.raise_for_status()
        return r.json()

    def get_lobby(self, lobby_id: str) -> Lobby:
        r = requests.get(f'{CROISSANT_BASE_URL}/lobbies/{lobby_id}')
        r.raise_for_status()
        return Lobby(**r.json())

    def get_my_lobby(self) -> Lobby:
        if not self.token:
            raise Exception('Token is required')
        r = requests.get(f'{CROISSANT_BASE_URL}/lobbies/user/@me', headers=self._headers())
        r.raise_for_status()
        return Lobby(**r.json())

    def get_user_lobby(self, user_id: str) -> Lobby:
        r = requests.get(f'{CROISSANT_BASE_URL}/lobbies/user/{user_id}')
        r.raise_for_status()
        return Lobby(**r.json())

    def join_lobby(self, lobby_id: str) -> dict:
        if not self.token:
            raise Exception('Token is required')
        r = requests.post(f'{CROISSANT_BASE_URL}/lobbies/{lobby_id}/join', headers=self._headers())
        r.raise_for_status()
        return r.json()

    def leave_lobby(self, lobby_id: str) -> dict:
        if not self.token:
            raise Exception('Token is required')
        r = requests.post(f'{CROISSANT_BASE_URL}/lobbies/{lobby_id}/leave', headers=self._headers())
        r.raise_for_status()
        return r.json()

    # --- STUDIOS ---
    def create_studio(self, studio_name: str) -> dict:
        if not self.token:
            raise Exception('Token is required')
        r = requests.post(f'{CROISSANT_BASE_URL}/studios', headers=self._headers(True), json={'studioName': studio_name})
        r.raise_for_status()
        return r.json()

    def get_studio(self, studio_id: str) -> Studio:
        r = requests.get(f'{CROISSANT_BASE_URL}/studios/{studio_id}')
        r.raise_for_status()
        return Studio(**r.json())

    def get_my_studios(self) -> List[Studio]:
        if not self.token:
            raise Exception('Token is required')
        r = requests.get(f'{CROISSANT_BASE_URL}/studios/user/@me', headers=self._headers())
        r.raise_for_status()
        return [Studio(**s) for s in r.json()]

    def add_user_to_studio(self, studio_id: str, user_id: str) -> dict:
        if not self.token:
            raise Exception('Token is required')
        r = requests.post(f'{CROISSANT_BASE_URL}/studios/{studio_id}/add-user', headers=self._headers(True), json={'userId': user_id})
        r.raise_for_status()
        return r.json()

    def remove_user_from_studio(self, studio_id: str, user_id: str) -> dict:
        if not self.token:
            raise Exception('Token is required')
        r = requests.post(f'{CROISSANT_BASE_URL}/studios/{studio_id}/remove-user', headers=self._headers(True), json={'userId': user_id})
        r.raise_for_status()
        return r.json()

    # --- TRADES ---
    def start_or_get_pending_trade(self, user_id: str) -> Trade:
        if not self.token:
            raise Exception('Token is required')
        r = requests.post(f'{CROISSANT_BASE_URL}/trades/start-or-latest/{user_id}', headers=self._headers())
        r.raise_for_status()
        return Trade(**r.json())

    def get_trade(self, trade_id: str) -> Trade:
        if not self.token:
            raise Exception('Token is required')
        r = requests.get(f'{CROISSANT_BASE_URL}/trades/{trade_id}', headers=self._headers())
        r.raise_for_status()
        return Trade(**r.json())

    def get_user_trades(self, user_id: str) -> List[Trade]:
        if not self.token:
            raise Exception('Token is required')
        r = requests.get(f'{CROISSANT_BASE_URL}/trades/user/{user_id}', headers=self._headers())
        r.raise_for_status()
        return [Trade(**t) for t in r.json()]

    def add_item_to_trade(self, trade_id: str, trade_item: dict) -> dict:
        if not self.token:
            raise Exception('Token is required')
        r = requests.post(f'{CROISSANT_BASE_URL}/trades/{trade_id}/add-item', headers=self._headers(True), json={'tradeItem': trade_item})
        r.raise_for_status()
        return r.json()

    def remove_item_from_trade(self, trade_id: str, trade_item: dict) -> dict:
        if not self.token:
            raise Exception('Token is required')
        r = requests.post(f'{CROISSANT_BASE_URL}/trades/{trade_id}/remove-item', headers=self._headers(True), json={'tradeItem': trade_item})
        r.raise_for_status()
        return r.json()

    def approve_trade(self, trade_id: str) -> dict:
        if not self.token:
            raise Exception('Token is required')
        r = requests.put(f'{CROISSANT_BASE_URL}/trades/{trade_id}/approve', headers=self._headers())
        r.raise_for_status()
        return r.json()

    def cancel_trade(self, trade_id: str) -> dict:
        if not self.token:
            raise Exception('Token is required')
        r = requests.put(f'{CROISSANT_BASE_URL}/trades/{trade_id}/cancel', headers=self._headers())
        r.raise_for_status()
        return r.json()

    # --- OAUTH2 ---
    def get_oauth2_app(self, client_id: str) -> OAuth2App:
        r = requests.get(f'{CROISSANT_BASE_URL}/oauth2/app/{client_id}')
        r.raise_for_status()
        return OAuth2App(**r.json())

    def create_oauth2_app(self, name: str, redirect_urls: List[str]) -> dict:
        if not self.token:
            raise Exception('Token is required')
        r = requests.post(f'{CROISSANT_BASE_URL}/oauth2/app', headers=self._headers(True), json={'name': name, 'redirect_urls': redirect_urls})
        r.raise_for_status()
        return r.json()

    def get_my_oauth2_apps(self) -> List[OAuth2App]:
        if not self.token:
            raise Exception('Token is required')
        r = requests.get(f'{CROISSANT_BASE_URL}/oauth2/apps', headers=self._headers())
        r.raise_for_status()
        return [OAuth2App(**a) for a in r.json()]

    def update_oauth2_app(self, client_id: str, data: dict) -> dict:
        if not self.token:
            raise Exception('Token is required')
        r = requests.patch(f'{CROISSANT_BASE_URL}/oauth2/app/{client_id}', headers=self._headers(True), json=data)
        r.raise_for_status()
        return r.json()

    def delete_oauth2_app(self, client_id: str) -> dict:
        if not self.token:
            raise Exception('Token is required')
        r = requests.delete(f'{CROISSANT_BASE_URL}/oauth2/app/{client_id}', headers=self._headers())
        r.raise_for_status()
        return r.json()

    def authorize(self, client_id: str, redirect_uri: str) -> dict:
        if not self.token:
            raise Exception('Token is required')
        r = requests.get(f'{CROISSANT_BASE_URL}/oauth2/authorize', headers=self._headers(), params={'client_id': client_id, 'redirect_uri': redirect_uri})
        r.raise_for_status()
        return r.json()

    def get_user_by_code(self, code: str, client_id: str) -> User:
        r = requests.get(f'{CROISSANT_BASE_URL}/oauth2/user', params={'code': code, 'client_id': client_id})
        r.raise_for_status()
        return User(**r.json())

File: public/downloadables/croissant_api.rb
require 'net/http'
require 'json'
require 'uri'

CROISSANT_BASE_URL = 'https://croissant-api.fr/api'

# Game represents a game in the Croissant API
class Game
  attr_accessor :game_id, :name, :description, :owner_id, :download_link, :price, :show_in_store,
                :icon_hash, :splash_hash, :banner_hash, :genre, :release_date, :developer,
                :publisher, :platforms, :rating, :website, :trailer_link, :multiplayer

  def initialize(attrs = {})
    attrs.each { |k, v| send("#{k}=", v) if respond_to?("#{k}=") }
  end
end

# User represents a user in the Croissant API
class User
  attr_accessor :user_id, :username, :email, :balance, :verified, :steam_id, :steam_username,
                :steam_avatar_url, :is_studio, :admin, :disabled, :google_id, :discord_id,
                :studios, :roles, :inventory, :owned_items, :created_games, :have_authenticator,
                :verification_key

  def initialize(attrs = {})
    attrs.each { |k, v| send("#{k}=", v) if respond_to?("#{k}=") }
  end
end

# Item represents an item in the Croissant API
class Item
  attr_accessor :item_id, :name, :description, :price, :owner, :show_in_store, :icon_hash, :deleted

  def initialize(attrs = {})
    attrs.each { |k, v| send("#{k}=", v) if respond_to?("#{k}=") }
  end
end

# InventoryItem represents an item in a user's inventory
class InventoryItem
  attr_accessor :item_id, :name, :description, :amount, :icon_hash

  def initialize(attrs = {})
    attrs.each { |k, v| send("#{k}=", v) if respond_to?("#{k}=") }
  end
end

# LobbyUser represents a user in a lobby
class LobbyUser
  attr_accessor :username, :user_id, :verified, :steam_username, :steam_avatar_url, :steam_id

  def initialize(attrs = {})
    attrs.each { |k, v| send("#{k}=", v) if respond_to?("#{k}=") }
  end
end

# Lobby represents a lobby
class Lobby
  attr_accessor :lobby_id, :users

  def initialize(attrs = {})
    @lobby_id = attrs['lobby_id']
    @users = (attrs['users'] || []).map { |u| LobbyUser.new(u) }
  end
end

# StudioUser represents a user in a studio
class StudioUser
  attr_accessor :user_id, :username, :verified, :admin

  def initialize(attrs = {})
    attrs.each { |k, v| send("#{k}=", v) if respond_to?("#{k}=") }
  end
end

# Studio represents a studio
class Studio
  attr_accessor :user_id, :username, :verified, :admin_id, :is_admin, :api_key, :users

  def initialize(attrs = {})
    @user_id = attrs['user_id']
    @username = attrs['username']
    @verified = attrs['verified']
    @admin_id = attrs['admin_id']
    @is_admin = attrs['is_admin']
    @api_key = attrs['api_key']
    @users = (attrs['users'] || []).map { |u| StudioUser.new(u) }
  end
end

# TradeItem represents a trade item
class TradeItem
  attr_accessor :item_id, :amount

  def initialize(attrs = {})
    attrs.each { |k, v| send("#{k}=", v) if respond_to?("#{k}=") }
  end
end

# TradeItemInfo represents enriched trade item information
class TradeItemInfo
  attr_accessor :item_id, :name, :description, :icon_hash, :amount

  def initialize(attrs = {})
    attrs.each { |k, v| send("#{k}=", v) if respond_to?("#{k}=") }
  end
end

# Trade represents a trade with enriched item information
class Trade
  attr_accessor :id, :from_user_id, :to_user_id, :from_user_items, :to_user_items,
                :approved_from_user, :approved_to_user, :status, :created_at, :updated_at

  def initialize(attrs = {})
    @id = attrs['id']
    @from_user_id = attrs['from_user_id']
    @to_user_id = attrs['to_user_id']
    @from_user_items = (attrs['from_user_items'] || []).map { |i| TradeItemInfo.new(i) }
    @to_user_items = (attrs['to_user_items'] || []).map { |i| TradeItemInfo.new(i) }
    @approved_from_user = attrs['approved_from_user']
    @approved_to_user = attrs['approved_to_user']
    @status = attrs['status']
    @created_at = attrs['created_at']
    @updated_at = attrs['updated_at']
  end
end

# SearchResult represents global search results
class SearchResult
  attr_accessor :users, :items, :games

  def initialize(attrs = {})
    @users = (attrs['users'] || []).map { |u| User.new(u) }
    @items = (attrs['items'] || []).map { |i| Item.new(i) }
    @games = (attrs['games'] || []).map { |g| Game.new(g) }
  end
end

# CroissantAPI provides methods to interact with the Croissant API
class CroissantAPI
  def initialize(token = nil)
    raise 'Token is required' if token.nil? || token.empty?
    @token = token
  end

  def get(path, auth: false)
    uri = URI("#{CROISSANT_BASE_URL}#{path}")
    req = Net::HTTP::Get.new(uri)
    req['Authorization'] = "Bearer #{@token}" if auth && @token
    res = Net::HTTP.start(uri.hostname, uri.port, use_ssl: true) { |http| http.request(req) }
    JSON.parse(res.body)
  end

  def post(path, body = {}, auth: false)
    uri = URI("#{CROISSANT_BASE_URL}#{path}")
    req = Net::HTTP::Post.new(uri)
    req['Content-Type'] = 'application/json'
    req['Authorization'] = "Bearer #{@token}" if auth && @token
    req.body = body.to_json
    res = Net::HTTP.start(uri.hostname, uri.port, use_ssl: true) { |http| http.request(req) }
    JSON.parse(res.body)
  end

  def put(path, body = {}, auth: false)
    uri = URI("#{CROISSANT_BASE_URL}#{path}")
    req = Net::HTTP::Put.new(uri)
    req['Content-Type'] = 'application/json'
    req['Authorization'] = "Bearer #{@token}" if auth && @token
    req.body = body.to_json
    res = Net::HTTP.start(uri.hostname, uri.port, use_ssl: true) { |http| http.request(req) }
    JSON.parse(res.body)
  end

  def delete(path, auth: false)
    uri = URI("#{CROISSANT_BASE_URL}#{path}")
    req = Net::HTTP::Delete.new(uri)
    req['Authorization'] = "Bearer #{@token}" if auth && @token
    res = Net::HTTP.start(uri.hostname, uri.port, use_ssl: true) { |http| http.request(req) }
    JSON.parse(res.body)
  end

  # --- USERS ---
  
  # Get the current authenticated user's profile, including studios, roles, inventory, owned items, and created games.
  def get_me
    User.new(get('/users/@me', auth: true))
  end

  # Get a user by userId. userId can be a Croissant ID, Discord ID, Google ID or Steam ID.
  def get_user(user_id)
    User.new(get("/users/#{user_id}"))
  end

  # Search for users by username.
  def search_users(query)
    get("/users/search?q=#{URI.encode_www_form_component(query)}").map { |u| User.new(u) }
  end

  # Check the verification key for the user.
  def verify_user(user_id, verification_key)
    post('/users/auth-verification', { userId: user_id, verificationKey: verification_key })
  end

  # Transfer credits from one user to another.
  def transfer_credits(target_user_id, amount)
    post('/users/transfer-credits', { targetUserId: target_user_id, amount: amount }, auth: true)
  end

  # Change username (authenticated user only).
  def change_username(username)
    post('/users/change-username', { username: username }, auth: true)
  end

  # Change password (authenticated user only).
  def change_password(old_password, new_password, confirm_password)
    post('/users/change-password', {
      oldPassword: old_password,
      newPassword: new_password,
      confirmPassword: confirm_password
    }, auth: true)
  end

  # --- GAMES ---
  
  # List all games visible in the store.
  def list_games
    get('/games').map { |g| Game.new(g) }
  end

  # Search for games by name, genre, or description.
  def search_games(query)
    get("/games/search?q=#{URI.encode_www_form_component(query)}").map { |g| Game.new(g) }
  end

  # Get a game by gameId.
  def get_game(game_id)
    Game.new(get("/games/#{game_id}"))
  end

  # Get all games created by the authenticated user.
  def get_my_created_games
    get('/games/@mine', auth: true).map { |g| Game.new(g) }
  end

  # Get all games owned by the authenticated user.
  def get_my_owned_games
    get('/games/list/@me', auth: true).map { |g| Game.new(g) }
  end

  # Create a new game.
  def create_game(game_hash)
    post('/games', game_hash, auth: true)
  end

  # Update an existing game.
  def update_game(game_id, game_hash)
    Game.new(put("/games/#{game_id}", game_hash, auth: true))
  end

  # Buy a game.
  def buy_game(game_id)
    post("/games/#{game_id}/buy", {}, auth: true)
  end

  # --- ITEMS ---
  
  # Get all non-deleted items visible in store.
  def list_items
    get('/items').map { |i| Item.new(i) }
  end

  # Get all items owned by the authenticated user.
  def get_my_items
    get('/items/@mine', auth: true).map { |i| Item.new(i) }
  end

  # Search for items by name (only those visible in store).
  def search_items(query)
    get("/items/search?q=#{URI.encode_www_form_component(query)}").map { |i| Item.new(i) }
  end

  # Get a single item by itemId.
  def get_item(item_id)
    Item.new(get("/items/#{item_id}"))
  end

  # Create a new item.
  def create_item(item_hash)
    post('/items/create', item_hash, auth: true)
  end

  # Update an existing item.
  def update_item(item_id, item_hash)
    put("/items/update/#{item_id}", item_hash, auth: true)
  end

  # Delete an item.
  def delete_item(item_id)
    delete("/items/delete/#{item_id}", auth: true)
  end

  # Buy an item.
  def buy_item(item_id, amount)
    post("/items/buy/#{item_id}", { amount: amount }, auth: true)
  end

  # Sell an item.
  def sell_item(item_id, amount)
    post("/items/sell/#{item_id}", { amount: amount }, auth: true)
  end

  # Give item occurrences to a user (owner only).
  def give_item(item_id, amount)
    post("/items/give/#{item_id}", { amount: amount }, auth: true)
  end

  # Consume item occurrences from a user (owner only).
  def consume_item(item_id, amount)
    post("/items/consume/#{item_id}", { amount: amount }, auth: true)
  end

  # Drop item occurrences from your inventory.
  def drop_item(item_id, amount)
    post("/items/drop/#{item_id}", { amount: amount }, auth: true)
  end

  # --- INVENTORY ---
  
  # Get the inventory of the authenticated user.
  def get_my_inventory
    get('/inventory/@me', auth: true).map { |i| InventoryItem.new(i) }
  end

  # Get the inventory of a user.
  def get_inventory(user_id)
    get("/inventory/#{user_id}").map { |i| InventoryItem.new(i) }
  end

  # --- LOBBIES ---
  
  # Create a new lobby.
  def create_lobby
    post('/lobbies', {}, auth: true)
  end

  # Get a lobby by lobbyId.
  def get_lobby(lobby_id)
    Lobby.new(get("/lobbies/#{lobby_id}"))
  end

  # Get the lobby the authenticated user is in.
  def get_my_lobby
    Lobby.new(get('/lobbies/user/@me', auth: true))
  end

  # Get the lobby a user is in.
  def get_user_lobby(user_id)
    Lobby.new(get("/lobbies/user/#{user_id}"))
  end

  # Join a lobby.
  def join_lobby(lobby_id)
    post("/lobbies/#{lobby_id}/join", {}, auth: true)
  end

  # Leave a lobby.
  def leave_lobby(lobby_id)
    post("/lobbies/#{lobby_id}/leave", {}, auth: true)
  end

  # --- STUDIOS ---
  
  # Create a new studio.
  def create_studio(studio_name)
    post('/studios', { studioName: studio_name }, auth: true)
  end

  # Get a studio by studioId.
  def get_studio(studio_id)
    Studio.new(get("/studios/#{studio_id}"))
  end

  # Get all studios the authenticated user is part of.
  def get_my_studios
    get('/studios/user/@me', auth: true).map { |s| Studio.new(s) }
  end

  # Add a user to a studio.
  def add_user_to_studio(studio_id, user_id)
    post("/studios/#{studio_id}/add-user", { userId: user_id }, auth: true)
  end

  # Remove a user from a studio.
  def remove_user_from_studio(studio_id, user_id)
    post("/studios/#{studio_id}/remove-user", { userId: user_id }, auth: true)
  end

  # --- TRADES ---
  
  # Start a new trade or get the latest pending trade with a user.
  def start_or_get_pending_trade(user_id)
    Trade.new(post("/trades/start-or-latest/#{user_id}", {}, auth: true))
  end

  # Get a trade by ID with enriched item information.
  def get_trade(trade_id)
    Trade.new(get("/trades/#{trade_id}", auth: true))
  end

  # Get all trades for a user with enriched item information.
  def get_my_trades
    get('/trades/user/@me', auth: true).map { |t| Trade.new(t) }
  end

  # Add an item to a trade.
  def add_item_to_trade(trade_id, trade_item)
    post("/trades/#{trade_id}/add-item", { tradeItem: trade_item }, auth: true)
  end

  # Remove an item from a trade.
  def remove_item_from_trade(trade_id, trade_item)
    post("/trades/#{trade_id}/remove-item", { tradeItem: trade_item }, auth: true)
  end

  # Approve a trade.
  def approve_trade(trade_id)
    put("/trades/#{trade_id}/approve", {}, auth: true)
  end

  # Cancel a trade.
  def cancel_trade(trade_id)
    put("/trades/#{trade_id}/cancel", {}, auth: true)
  end

  # --- SEARCH ---
  
  # Global search across users, items, and games.
  def global_search(query)
    SearchResult.new(get("/search?q=#{URI.encode_www_form_component(query)}"))
  end

  # --- OAUTH2 ---

  # Get an OAuth2 app by its client ID.
  def get_oauth2_app(client_id)
    get("/oauth2/app/#{client_id}", auth: false)
  end

  # Create a new OAuth2 app.
  def create_oauth2_app(name, redirect_urls)
    post('/oauth2/app', { name: name, redirect_urls: redirect_urls }, auth: true)
  end

  # Get all OAuth2 apps owned by the authenticated user.
  def get_my_oauth2_apps
    get('/oauth2/apps', auth: true)
  end

  # Update an OAuth2 app.
  def update_oauth2_app(client_id, data)
    uri = "/oauth2/app/#{client_id}"
    Net::HTTP.start(URI(CROISSANT_BASE_URL + uri).hostname, URI(CROISSANT_BASE_URL + uri).port, use_ssl: true) do |http|
      req = Net::HTTP::Patch.new(URI(CROISSANT_BASE_URL + uri))
      req['Content-Type'] = 'application/json'
      req['Authorization'] = "Bearer #{@token}"
      req.body = data.to_json
      res = http.request(req)
      JSON.parse(res.body)
    end
  end

  # Delete an OAuth2 app by its client ID.
  def delete_oauth2_app(client_id)
    delete("/oauth2/app/#{client_id}", auth: true)
  end

  # Authorize a user for an OAuth2 app.
  def authorize_oauth2(client_id, redirect_uri)
    get("/oauth2/authorize?client_id=#{URI.encode_www_form_component(client_id)}&redirect_uri=#{URI.encode_www_form_component(redirect_uri)}", auth: true)
  end

  # Get a user by OAuth2 code and client ID.
  def get_user_by_code(code, client_id)
    get("/oauth2/user?code=#{URI.encode_www_form_component(code)}&client_id=#{URI.encode_www_form_component(client_id)}")
  end

  # --- ITEMS METADATA ---

  # Give an item to a user with metadata.
  def give_item_with_metadata(item_id, amount, user_id, metadata = nil)
    body = { amount: amount, userId: user_id }
    body[:metadata] = metadata if metadata
    post("/items/give/#{item_id}", body, auth: true)
  end

  # Update metadata for an item instance.
  def update_item_metadata(item_id, unique_id, metadata)
    put("/items/update-metadata/#{item_id}", { uniqueId: unique_id, metadata: metadata }, auth: true)
  end

  # Consume an item instance (with params hash).
  def consume_item_instance(item_id, params)
    post("/items/consume/#{item_id}", params, auth: true)
  end

  # Drop an item instance (with params hash).
  def drop_item_instance(item_id, params)
    post("/items/drop/#{item_id}", params, auth: true)
  end
end
